components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java

```diff
@@ -667,6 +678,72 @@ private static void addUserAttributesToCache(AccessTokenDO accessTokenBean,
         }
     }
 
+    private void executePreIssueAccessTokenActions(RefreshTokenValidationDataDO refreshTokenValidationDataDO,
+                                                   OAuthTokenReqMessageContext tokenReqMessageContext)
+            throws IdentityOAuth2Exception {
+
+        if (checkExecutePreIssueAccessTokensActions(refreshTokenValidationDataDO, tokenReqMessageContext)) {
+
+            setCustomizedAccessTokenAttributesToMessageContext(refreshTokenValidationDataDO, tokenReqMessageContext);
+
+            Map<String, Object> additionalProperties = new HashMap<>();
+            Consumer<Map<String, Object>> mapInitializer =
+                    map -> map.put("tokenMessageContext", tokenReqMessageContext);
+            mapInitializer.accept(additionalProperties);
+
+            try {
+                ActionExecutionStatus executionStatus =
+                        OAuthComponentServiceHolder.getInstance().getActionExecutorService()
+                                .execute(ActionType.PRE_ISSUE_ACCESS_TOKEN, additionalProperties,
+                                        IdentityTenantUtil.getTenantDomain(IdentityTenantUtil.getLoginTenantId()));
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format(
+                            "Invoked pre issue access token action for clientID: %s grant types: %s. Status: %s",
+                            tokenReqMessageContext.getOauth2AccessTokenReqDTO().getClientId(),
+                            tokenReqMessageContext.getOauth2AccessTokenReqDTO().getGrantType(),
+                            Optional.ofNullable(executionStatus).isPresent() ? executionStatus.getStatus() : "NA"));
+                }
+            } catch (ActionExecutionException e) {
+                // If error ignore and proceed
+                log.error("Error while executing pre issue access token action", e);
+            }
+        }
+    }
+
+    private boolean checkExecutePreIssueAccessTokensActions(RefreshTokenValidationDataDO refreshTokenValidationDataDO,
+                                                            OAuthTokenReqMessageContext tokenReqMessageContext)
+            throws IdentityOAuth2Exception {
+
+        OAuthAppDO oAuthAppBean = getOAuthApp(tokenReqMessageContext.getOauth2AccessTokenReqDTO().getClientId());
+        String grantType = refreshTokenValidationDataDO.getGrantType();
+
+        // Allow if refresh token is issued for token requests from following grant types and,
+        // for JWT access tokens only.
+        return (OAuthConstants.GrantTypes.AUTHORIZATION_CODE.equals(grantType) ||
```

Comment:
Any reason to stick to this particular grant types only?

Reply:
This delivery focuses on these grant types only.

So a specific check was added to make sure this gets engaged for tested grant types and nothing breaks if it's a different grant type.